csharp/test/Drivers/Interop/Snowflake/CastTests.cs

```diff
@@ -0,0 +1,397 @@
+            Console.WriteLine(insertNumberStatement);
+            _statement.SqlQuery = insertNumberStatement;
+            UpdateResult updateResult = _statement.ExecuteUpdate();
+            Assert.Equal(expectedAffectedRows, updateResult.AffectedRows);
+        }
+
+        private async void SelectWithCastAndValidateValue(
+            string table,
+            string castFunction,
+            string castExpression,
+            object value,
+            ArrowTypeId expectedType)
+        {
+            QueryResult queryResult = PerformQuery(table, castFunction, castExpression);
+            await ValidateCast(value, expectedType, queryResult);
+        }
+
+        private void SelectWithCastAndValidateException(
+            string table,
+            string castFunction,
+            string castExpression,
+            Type expectedExceptionType,
+            string[] expectedExceptionTextContains = null)
+        {
+            Exception actualException = Assert.Throws(expectedExceptionType, () => PerformQuery(table, castFunction, castExpression));
+            AssertContainsAll(expectedExceptionTextContains, actualException.Message);
+        }
+
+        private static void AssertContainsAll(string[] expectedTexts, string actualText)
+        {
+            if (expectedTexts == null) { return; };
+            foreach (string text in expectedTexts)
+            {
+                Assert.Contains(text, actualText);
+            }
+        }
+
+        private QueryResult PerformQuery(string table, string castFunction, string castExpression)
+        {
+            string selectStatement = string.Format(
+                "SELECT {0}({1}) AS CASTRESULT FROM {2};",
+                castFunction,
+                castExpression,
+                table);
+            Console.WriteLine(selectStatement);
+            _statement.SqlQuery = selectStatement;
+            return _statement.ExecuteQuery();
+        }
+
+        private static async Task ValidateCast(object value, ArrowTypeId expectedType, QueryResult queryResult)
+        {
+            Assert.Equal(1, queryResult.RowCount);
+            using IArrowArrayStream stream = queryResult.Stream;
+            Field field = stream.Schema.GetFieldByName("CASTRESULT");
+            Assert.NotNull(field);
+            while (true)
+            {
+                using RecordBatch nextBatch = await stream.ReadNextRecordBatchAsync();
+                if (nextBatch == null) { break; }
+                Assert.Equal(expectedType, field.DataType.TypeId);
+                IArrowArray valueArray = nextBatch.Column(0);
+
+                switch (field.DataType.TypeId)
+                {
+                    case ArrowTypeId.Double:
+                        var doubleArray = valueArray as DoubleArray;
+                        for (int i = 0; i < doubleArray.Length; i++)
+                        {
+                            Assert.Equal(Convert.ToDouble(value), doubleArray.GetValue(i));
```

Comment:
Suggestion, add a custom error message to these generic assertions to provide traceability if there's a failure (which test value is failing, which index).

Reply:
Hmm. The only problem the Xunit framework don't have any overloads for adding a custom message (known issue: https://github.com/xunit/xunit/issues/350). What I can say is 1. We only every have one value to test; 2. the exception messages thrown are pretty specific with full stack trace and runtime values.